### ISSUE TITLE
grid_buysell_paper.py

### ISSUE THREAD

**Issue by Holender26** [Holender26/Scalp-bot#4]
cat > grid_buysell_paper.py <<'PY'
import json, time, os, csv, math
from datetime import datetime, timezone
import argparse
import requests

BINANCE_PRICE_URL = "/api/v3/ticker/price"

def now_utc():
    return datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S")

def load_config(path):
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)

def to_binance_symbol(symbol: str) -> str:
    return symbol.replace("/", "").upper().strip()

def get_price(symbol_binance: str, base_url: str) -> float:
    url = base_url.rstrip("/") + BINANCE_PRICE_URL
    r = requests.get(url, params={"symbol": symbol_binance}, timeout=10)
    r.raise_for_status()
    return float(r.json()["price"])

def ensure_csv(path: str, header):
    if os.path.exists(path):
        return
    with open(path, "w", newline="", encoding="utf-8") as f:
        csv.writer(f).writerow(header)

def append_csv(path: str, row):
    with open(path, "a", newline="", encoding="utf-8") as f:
        csv.writer(f).writerow(row)

def pick_active_strategy(strategies: dict) -> str:
    for name, cfg in strategies.items():
        if isinstance(cfg, dict) and cfg.get("enabled") is True:
            return name
    return "none"

def round_down(x, step):
    return math.floor(x / step) * step

def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--config", required=True)
    args = ap.parse_args()

    cfg = load_config(args.config)
    if cfg.get("mode") != "paper":
        print("This script is PAPER only. Set mode=paper in config.")
        return

    ex = cfg.get("exchange", {})
    base_url = ex.get("base_url", "https://api.binance.com")

    market = cfg.get("market", {})
    symbol = market.get("symbol", "BTC/USDT")
    symbol_b = to_binance_symbol(symbol)

    risk = cfg.get("risk", {})
    max_daily_loss_pct = float(risk.get("max_daily_loss_pct", 5.0))
    cooldown_after_loss = int(risk.get("cooldown_seconds_after_loss", 120))

    exec_cfg = cfg.get("execution", {})
    slippage_bps = float(exec_cfg.get("slippage_bps", 0))
    fee_bps = float(exec_cfg.get("fee_bps", 0))
    cost_bps = slippage_bps + fee_bps

    strategies = cfg.get("strategies", {})
    active = pick_active_strategy(strategies)
    if active != "grid":
        print(f"Active strategy is '{active}'. Enable strategies.grid.enabled=true in config.")
        return

    grid = strategies.get("grid", {})
    spacing_pct = float(grid.get("grid_spacing_pct", 0.6)) / 100.0
    levels = int(grid.get("levels", 14))
    global_stop_pct = float(grid.get("global_stop_pct", 6.0)) / 100.0

    logging = cfg.get("logging", {})
    trades_csv = logging.get("trades_csv", "trades_12345.csv")
    events_csv = os.path.splitext(trades_csv)[0] + "_events.csv"

    ensure_csv(trades_csv, ["ts_utc","symbol","action","price","qty","quote_after","base_after","pnl_quote","note"])
    ensure_csv(events_csv, ["ts_utc","symbol","price","note"])

    # PAPER balances
    quote_start = 1000.0   # USDT start (możemy potem podpiąć pod config)
    base_start  = 0.0      # BTC start
    quote = quote_start
    base  = base_start

    # PnL dzienny (w USDT)
    day_key = datetime.now(timezone.utc).strftime("%Y-%m-%d")
    daily_pnl = 0.0
    cooldown_until = 0

    # Cena referencyjna i krok
    px0 = get_price(symbol_b, base_url)
    ref = px0
    step = ref * spacing_pct

    # Wielkość pojedynczego zlecenia (notional w USDT)
    # Prosto: 0.5% kapitału na order, min 10 USDT
    def order_notional():
        return max(10.0, quote_start * 0.005)

    # Open orders: list of dict {side, price, qty, link_price}
    orders = []

    def place_initial_grid():
        orders.clear()
        # BUY levels below ref
        for i in range(1, levels + 1):
            p = ref - i * step
            n = order_notional()
            qty = n / p
            orders.append({"side":"BUY", "price": round(p,2), "qty": qty})
        # SELL levels above ref (na start sprzedajemy tylko jeśli mamy base; jeśli nie mamy, i tak postawimy “virtual” sell,
        # ale będzie się fillował dopiero gdy base będzie dostępne)
        for i in range(1, levels + 1):
            p = ref + i * step
            n = order_notional()
            qty = n / p
            orders.append({"side":"SELL","price": round(p,2), "qty": qty})
        # sort: BUY malejąco (najbliższy), SELL rosnąco (najbliższy)
        # trzymamy w jednej liście, a przy fillach filtrujemy
        append_csv(events_csv, [now_utc(), symbol_b, px0, f"grid_set ref={ref:.2f} step={step:.2f} levels={levels}"])
    place_initial_grid()

    def est_cost_multiplier(side: str):
        # koszt wejścia/wyjścia: przy BUY płacisz więcej, przy SELL dostajesz mniej
        if side == "BUY":
            return 1 + cost_bps / 10000.0
        return 1 - cost_bps / 10000.0

    def do_fill(order, mkt_price):
        nonlocal quote, base, daily_pnl, cooldown_until
        side = order["side"]
        p = float(order["price"])
        qty = float(order["qty"])

        # wykonanie po "p" z kosztem
        if side == "BUY":
            exec_p = p * est_cost_multiplier("BUY")
            cost_quote = exec_p * qty
            if quote < cost_quote:
                return False, "no_quote"
            quote_before = quote
            base_before = base
            quote -= cost_quote
            base  += qty
            # PnL realizuje się dopiero przy SELL — tu zapisujemy tylko fill
            append_csv(trades_csv, [now_utc(), symbol_b, "BUY_FILL", round(exec_p,2), round(qty,8),
                                    round(quote,4), round(base,8), 0.0, f"spent={cost_quote:.4f}"])
            return True, f"buy filled (quote {quote_before:.2f}->{quote:.2f}, base {base_before:.6f}->{base:.6f})"

        else:  # SELL
            exec_p = p * est_cost_multiplier("SELL")
            proceeds = exec_p * qty
            if base < qty:
                return False, "no_base"
            quote_before = quote
            base_before = base
            base  -= qty
            quote += proceeds
            # Realizowany PnL liczymy jako “zysk w quote” względem wartości qty po cenie ref? (uprościmy: zysk=proceeds - cost_basis)
            # Żeby nie udawać księgowości, liczymy PnL dzienny jako delta quote względem poprzedniego stanu po SELL.
            pnl = quote - quote_before
            daily_pnl += pnl
            append_csv(trades_csv, [now_utc(), symbol_b, "SELL_FILL", round(exec_p,2), round(qty,8),
                                    round(quote,4), round(base,8), round(pnl,4), f"got={proceeds:.4f}"])
            return True, f"sell filled (quote {quote_before:.2f}->{quote:.2f}, base {base_before:.6f}->{base:.6f})"

    def replace_opposite(order):
        # klasyczny grid: po BUY stawiamy SELL o 1 krok wyżej; po SELL stawiamy BUY o 1 krok niżej
        side = order["side"]
        p = float(order["price"])
        qty = float(order["qty"])
        if side == "BUY":
            new_p = p + step
            orders.append({"side":"SELL","price": round(new_p,2), "qty": qty})
        else:
            new_p = p - step
            orders.append({"side":"BUY","price": round(new_p,2), "qty": qty})

    print(f"[{now_utc()}] START PAPER GRID BUY+SELL id={cfg.get('id')} symbol={symbol}({symbol_b})")
    print(f"[{now_utc()}] ref={ref:.2f} step={step:.2f} levels={levels} cost_bps={cost_bps}")
    print(f"[{now_utc()}] balances: quote={quote:.2f} base={base:.6f}")

    tick = 0
    while True:
        try:
            # dzienny reset
            day_now = datetime.now(timezone.utc).strftime("%Y-%m-%d")
            if day_now != day_key:
                day_key = day_now
                daily_pnl = 0.0
                cooldown_until = 0
                append_csv(events_csv, [now_utc(), symbol_b, "", "daily_reset"])

            px = get_price(symbol_b, base_url)
            tick += 1

            if tick % 2 == 0:
                append_csv(events_csv, [now_utc(), symbol_b, px, f"feed orders={len(orders)} quote={quote:.2f} base={base:.6f}"])

            # dzienny limit straty (liczony po SELL fillach w uproszczeniu)
            if daily_pnl <= -abs(quote_start) * (max_daily_loss_pct / 100.0):
                append_csv(events_csv, [now_utc(), symbol_b, px, f"HALT daily_loss_limit pnl={daily_pnl:.2f}"])
                print(f"[{now_utc()}] HALT: daily loss limit hit. pnl={daily_pnl:.2f}")
                time.sleep(5)
                continue

            if int(time.time()) < cooldown_until:
                time.sleep(2)
                continue

            # global stop: jeśli cena odjedzie mocno od ref -> pauza i recenter (bez “panic sell”)
            if px <= ref * (1 - global_stop_pct) or px >= ref * (1 + global_stop_pct):
                append_csv(events_csv, [now_utc(), symbol_b, px, f"GLOBAL_STOP_RECENTER ref={ref:.2f}"])
                ref = px
                step = ref * spacing_pct
                place_initial_grid()
                time.sleep(2)
                continue

            # Fill logic:
            # - BUY fills when px <= buy_price
            # - SELL fills when px >= sell_price
            filled_any = False

            # sort "closest first" each loop for stable behavior
            buys = sorted([o for o in orders if o["side"]=="BUY"], key=lambda x: -x["price"])   # highest buy first
            sells = sorted([o for o in orders if o["side"]=="SELL"], key=lambda x: x["price"]) # lowest sell first

            # próbuj wypełnić kilka orderów, ale z limitem na tick (żeby nie robić “lawiny”)
            max_fills_per_tick = 3
            fills = 0

            # BUY fills
            for o in buys:
                if fills >= max_fills_per_tick:
                    break
                if px <= o["price"]:
                    ok, reason = do_fill(o, px)
                    if ok:
                        orders.remove(o)
                        replace_opposite(o)
                        filled_any = True
                        fills += 1
                    else:
                        # brak USDT -> nie fillujemy, zostawiamy
                        pass

            # SELL fills
            for o in sells:
                if fills >= max_fills_per_tick:
                    break
                if px >= o["price"]:
                    ok, reason = do_fill(o, px)
                    if ok:
                        orders.remove(o)
                        replace_opposite(o)
                        filled_any = True
                        fills += 1
                    else:
                        # brak BTC -> nie fillujemy, zostawiamy “wirtualnie”
                        pass

            if filled_any:
                append_csv(events_csv, [now_utc(), symbol_b, px, f"fills={fills} quote={quote:.2f} base={base:.6f} pnl_day={daily_pnl:.2f}"])

            time.sleep(2)

        except KeyboardInterrupt:
            print("\nStopped (CTRL+C).")
            break
        except Exception as e:
            append_csv(events_csv, [now_utc(), symbol_b, "", f"ERROR {e}"])
            time.sleep(3)

if __name__ == "__main__":
    main()
PY
